common/crypto/types.cpp

```diff
@@ -0,0 +1,85 @@
+/* Copyright 2018 Intel Corporation
```

Comment:
Can't we reuse the existing one from pdo (or at least refactor it)?  Or alternatively (and maybe better?)  we already "import" base64 (and hex_string) via pdo, so we could essentially dump the FPC version and just re-use the PDO version?

Reply:
> Can't we reuse the existing one from pdo (or at least refactor it)?
No, because PDO's version uses PDO's base64 package which is different.

> Or alternatively (and maybe better?) we already "import" base64 (and hex_string) via pdo, so we could essentially dump the FPC version and just re-use the PDO version?
We are not importing base64 from PDO!!! -- that's the point.
This `types.cpp` has been changed to use FPC's base64.

Let's clarify the problem. For performance reasons, PDO has/uses a modified version of the original base64 package. FPC still uses the original package. As the PDO crypto library requires `types.cpp` (which is the one using the base64 package), this has to be adapted accordingly for FPC. Most likely, in the future, we will adopt PDO's base64 package and dump the original one. This effort is not considered worthwhile, for what we are doing/need now.

Reply:
the crypto itself does not require the base64 functions from types.cpp; so one could easily address it by splitting types.cpp into two?  Also if there is such a performance difference that we changed it in PDO, why wouldn't we want it here?

Reply:
> the crypto itself does not require the base64 functions from types.cpp; so one could easily address it by splitting types.cpp into two?

Possibly. But since the file has some functionality that might be useful in FPC, it seems premature to do that.
Also, if FPC adopt PDO's base64 package, we need to neither split nor copy/modify `types.cpp`.

> Also if there is such a performance difference that we changed it in PDO, why wouldn't we want it here?

As I said, we will "most likely" adopt that.
Reasons not to do it now:
* personally, I have not run any related performance test
* adopting that means
    * either keeping 2 base64 packages, which is ugly
    * or dropping FPC's base64 package
        * in the latter case, you have modifications to do all around FPC -- which is reasonable, but in a separate PR.

Reply:
Hmm, but if you split types.cpp in two, you still can adopt the use in fpc later if we ever want to do so but also have a clean interface?  I have no strong feeling of going either the split-types.cpp or the one-base64-in-fpc route, i just am not excited in the version with the duplicated types.cpp.   btw: marcela probably should also weigh in on this for her graphene use-case?

Reply:
> Hmm, but if you split types.cpp in two, you still can adopt the use in fpc later if we ever want to do so but also have a clean interface? I have no strong feeling of going either the split-types.cpp or the one-base64-in-fpc route,

I see the split really premature, since in the future I believe we will adopt the PDO's base64.

> i just am not excited in the version with the duplicated types.cpp. btw: marcela probably should also weigh in on this for her graphene use-case?

Let me provide more reasons why this is not a problem.
1) Even base64 is duplicated -- we could have it as a submodule rather than shipping it.
2) The PDO submodule will not be "master". It will be tagged with a commit id, so the code will not change.
3) I think nobody has reasons to change the interfaces of crypto, types, FPC's base64 and PDO's base64. Therefore, removing or splitting `types.cpp` can done seamlessly in the future by only updating (c)makefiles without touching any code.

Reply:
i don't the split would have been premature. Actually to the contrary in hindsight: it would provide a clean interface to pdo crypto (which doesn't need base64 but the rest of types.cpp). But given that Mic now merged -- i guess i shouldn't have initially approved it before more of the questions popped up ..;-) --- it is now kindof a moot question.

I still don't like the types.cpp as-is but i can understand the resistance to switch right away to pdo base64.  What i would think, though, that in this case we have to add a comment in the FPC types.cpp saying that it is "modified clone" of `pdo/common/types.cpp` and why it's there.  And maybe add a TODO to revisit eventually whether we might want to eventually revisit the use of the more efficient PDO base64 variation.

Reply:
I still disagree on the split: I don't see why splitting `types` based on the base64 dependency would be cleaner. 

* ok for the comment
* ok for the todo to revisit it
* ok for checking the base64 variation

Reply:
just in case it was not clear: i was talking splitting types in PDO (and change pdo crypto to use only the base64-less version of it), not here. In that case you wouldn't need this types.cpp here anymore and could use the pdo crypto version right away?  Anyway, seems now a moot point ...